src/serde/tree_cache.rs

```diff
@@ -0,0 +1,744 @@
+use super::bytes32::{hash_blobs, Bytes32};
+use super::{ChildPos, PathBuilder};
+use crate::allocator::{Allocator, NodePtr, SExp};
+use crate::serde::serialized_length_atom;
+use crate::serde::RandomState;
+use crate::serde::VisitedNodes;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+
+const MIN_SERIALIZED_LENGTH: u64 = 4;
+
+#[derive(Clone, Debug)]
+struct NodeEntry {
+    /// the tree hash of this node. It may be None if it or any of its children
+    /// is the sentinel node, which means we can't compute the tree hash.
+    tree_hash: Option<Bytes32>,
+    /// a node can have an arbitrary number of parents, since they can be reused
+    /// this is a list of parent nodes, followed by whether we're the left or
+    /// right child. The u32 is an index into the node_entry vector.
+    parents: Vec<(u32, ChildPos)>,
+    /// if this node doesn't have a tree_hash, the serialized length is not
+    /// valid as it cannot be computed.
+    serialized_length: u64,
+    /// set to non-zero, if this node is pushed onto the parse stack. Since a
+    /// node can be on the stack multiple times, it's counted up every time it's
+    /// pushed, and counted down every time it's popped.
+    pub on_stack: u32,
+}
+
+struct PartialPath {
+    // the path we've built so far
+    path: PathBuilder,
+    // if we're traversing the stack, this is the stack position. Note that this
+    // is not an index into the stack array, it's a counter of how far away from
+    // the top of the stack we are. 0 means we're at the top, and we've found
+    // a path.
+    // -1 means we're not traversing the stack.
+    // TODO: these fields are mutually exclusive (stack_pos vs. idx and child).
+    // it might be nice to use an enum, but it's not obvious that it's simpler
+    // or faster
+    stack_pos: i32,
+    // if we're not traversing the stack, this is the next parent, and whether
+    // we're coming from the left or right child
+    idx: u32,
+    child: ChildPos,
+}
+
+enum CacheOp {
+    // node to traverse
+    Traverse(NodePtr),
+    // pair node
+    Cons(NodePtr),
+}
+
+#[derive(Clone)]
+pub struct TreeUndoState {
+    stack: Vec<u32>,
+    lookup: VisitedNodes,
+    sentinel_entry: Option<u32>,
+}
+
+#[derive(Default)]
+pub struct TreeCache {
+    /// caches extra metadata about a tree of nodes. The value is an index into
+    /// the node_entry vector.
+    cache: HashMap<NodePtr, u32>,
```

Comment:
Maybe `node_ptr_to_node_entry_index_map`? Kind of long though. We almost need a separate type that corresponds to indices into `node_entry` (just a specialization of `usize` or `u32`).

`node_ptr_2_entry_index`? The problem with `cache` is every time I see it, I have to remind myself what it is.

Reply:
maybe `node_map`? I think it would be a pretty big improvement to use a strong type for `u32` as well

Reply:
I agree that `cache` is not a good name. nothing about this is cache-like.